### DISCUSSION TITLE
Using Clocks in Sparta

### DISCUSSION THREAD

**Discussion by thheil-dmatrix** [sparcians/map#475] [General]
Greetings,

We are setting up new internal models using Sparta.  One thing that is confusing us is the Clock facility in Sparta.  Can we get some getting started guidance?  They appear to have all the features we need, but I haven't been able to actually instantiate and assign them as the "default" clock to our models.

I'm building clocks based on parameters in the model, in the model class constructor.  I can create the clock like so:

      clock_ = cm.makeClock("cons_clock", cm.getRoot(), 10, 6);

(The "10" and "6" would normally be parameters).

However, if I attempt to set them as the clock for the model's resource, as so:

      node->setClock(clock_.get());

we get an error:

     terminate called after throwing an instance of 'sparta::SpartaException'
        what():  Cannot set clock for node top.the_consumer once in the TREE_FINALIZING phase

That setClock call seems to be the way that schedule(...) type operations get a hold of the default clock to use to interpret scheduling delays.  That's what we are after.

The model constructor is the soonest we have access to the model parameters (as far as we know), and the soonest we can run model-specific code (against, as far as we know).  So I don't understand how we can setup model-specific clocks any earlier than this.  

What is the expected process or intended usage model here?  I've seen the documentation on Clocking that I've been able to find, but details are still sketchy for me.

Regards,

Timothy

**Comment by nitheshk1** [comment]
Hi @klingaard  appreciate your help answering our questions!

**Comment by klingaard** [comment]
Hi Timothy, appreciate the questions!

I agree this part of the simulation infrastructure is not very clear.  I'll use this discussion to shed light on it and hopefully find the time to properly document it.

First, you are correct in understanding that the Clock's main purpose is to be a "calculator" for event scheduling.  This is especially critical for disparate clock domains like a core talking to a memory subsystem via `Sync[In,Out]Ports`.  In addition, the Clock can help the modeler convert from Ticks (ps) to Cycles and vice-versa. 

If you're interested in just 1 clock domain, you don't need to do anything with ClockManager or perform any clock creation.  Just work under the assumption that 1 ps == 1 cycle and move on.  This is how the example CoreModel works as well as the [RISC-V perf model Olympia.
](https://github.com/riscv-software-src/riscv-perf-model)

But that's not your use case :grin:.  

For your use case, I'll assume something simple: that you need to create two clocks, each on its own frequency.  In this case you use the [`ClockManager`](https://sparcians.github.io/map/classsparta_1_1ClockManager.html) to create those clocks so that the framework has an understanding of the relationship between them.  From there, you must register those clocks (or set them) in `TreeNode` associated with the resource that will use each `Clock`.   This step must be done in the either the build phase or the subsequent configure phase of simulation setup (see Slides 10 - 18 on the [Sparta Overview](https://docs.google.com/presentation/d/e/2PACX-1vS1BWtVv0x3qXKQWAeECe2gsF9cMG3Zp2HnXJw52grCAcl21lv3a9pLW6J0lZ32e5DWdZkFyUMcE_AI/pub?start=false&loop=false&delayms=3000#slide=id.g7c850bbd55_1_60) slide deck).  An example of how to do that can be found in the [SyncPort_test.cpp](https://github.com/sparcians/map/blob/64e119f0321ca144b79071592513baedf028f958/sparta/test/SyncPort/SyncPort_test.cpp#L354-L356).

Clocks that are associated with a `TreeNode` are now the "master" clock or parent clock of all `TreeNode`s in the hierarchy:
```
       A   (has a clock_a)
     /   \
    B     C  <-- (TreeNode B and C use clock_a pulled from TreeNode A)
```
Because of this rule you cannot set the clock during the finalizing (or construction) phase of simulation.  Specifically, you cannot set the clock in the constructor of your Unit or Resource because it's quite possible that the sub-components (`B` & `C`) in the tree below that unit have already been initialized and grabbed what it thought was the correct clock to use.

But you'd like to parameterize the clocks...

Clocks can be created and associated with a `TreeNode` up to and including configuration stage.  In a simulation's configuration stage, you can grab any parameter from the tree https://github.com/sparcians/map/blob/64e119f0321ca144b79071592513baedf028f958/sparta/example/CoreModel/src/ExampleSimulation.cpp#L748-L749 and read its value.  In your use case, you can create clock frequencies or clock ratios in either parameter form or as an extension (there are [examples in the CoreModel](https://github.com/sparcians/map/blob/master/sparta/example/CoreModel/config_with_tree_node_extensions.yaml) of extensions).  

For example (using extensions):
```
void MySimulator::configureTree_()
{
    const auto clock_freq_extensions = root_node->getExtension("clock_freqs");
    const auto clock_freq_params = sparta::notNull(clock_freq_extensions)->getParameters();
    const auto cpu_clock_freq = sparta::notNull(clock_freq_params)->getParameter("cpu_clock_freq")->getValueAs<double>();
    const auto mss_clock_freq = sparta::notNull(clock_freq_params)->getParameter("mss_clock_freq")->getValueAs<double>();

    // ClockManager code...

    root_tn->getChild("cpu")->setClock(core_clock);
    root_tn->getChild("mss")->setClock(mss_clock);
}
```
Hope this helps.

**Reply by thheil-dmatrix** [reply]
Hi Knute,

Thanks for the detailed reply.  I coded based on that yesterday, and that got us where we need to be.  We ended up creating and configuring clocks separately from models, which made things more flexible if anything.

I close this issue.  Appreciate the prompt support.

Timothy

**Reply by klingaard** [reply]
Great!  Happy to help.
